relay-server/src/actors/upstream.rs

```diff
@@ -462,55 +612,83 @@ impl SendRequest {
 }
 
-impl<B, R, T: 'static, E: 'static> Handler<SendRequest<B, R>> for UpstreamRelay
+/// SendRequest<B> messages represent external messages that need to be sent to the upstream server
+/// and do not use Relay authentication.
+///
+/// The handler adds the message to one of the message queues and tries to advance the processing
+/// by sending a `PumpMessageQueue`.
+impl<B> Handler<SendRequest<B>> for UpstreamRelay
 where
-    B: RequestBuilder,
-    R: ResponseTransformer,
-    R::Result: IntoFuture<Item = T, Error = E>,
-    T: Send,
-    E: From<UpstreamRequestError> + Send,
+    B: RequestBuilder + Send,
 {
-    type Result = ResponseFuture<T, E>;
+    type Result = ResponseFuture<(), UpstreamRequestError>;
 
-    fn handle(&mut self, message: SendRequest<B, R>, _ctx: &mut Self::Context) -> Self::Result {
+    fn handle(&mut self, message: SendRequest<B>, ctx: &mut Self::Context) -> Self::Result {
         let SendRequest {
             method,
             path,
             builder,
-            transformer,
         } = message;
 
-        Box::new(
-            self.send_request(method, path, |b| builder.build_request(b))
-                .from_err()
-                .and_then(|r| transformer.transform_response(r)),
-        )
+        let ret_val = Box::new(
+            self.send_request(RequestPriority::Low, method, path, |b| {
+                builder.build_request(b)
+            })
+            .and_then(|client_response| {
+                client_response
+                    .payload()
+                    .for_each(|_| Ok(()))
+                    .map_err(UpstreamRequestError::PayloadFailed)
+            }),
+        );
+
+        ctx.notify(PumpHttpMessageQueue);
+        ret_val
     }
 }
 
-pub trait UpstreamQuery: Serialize {
+/// This handler handles messages sent on the mpsc channel that mark the end of an http request
+/// future. The handler decrements the counter of in-flight HTTP requests (since one was just
+/// finished) and tries to pump the http message queue by sending a `PumpHttpMessageQueue`
+///
+/// Every future representing an HTTP message sent by the ClientConnector is wrapped so that when
+/// it finishes or it is dropped a message is sent into a mpsc channel. This handler deals with
+/// the receiver end of the mpsc channel.
+///
+/// **Note:** An alternative, simpler, implementation would have been to increment the in-flight
+/// requests counter just before sending an http message and to decrement it when the future
+/// representing the sent message completes (on the future .then() method).
+/// While this approach would have simplified the design, no need for wrapping, no need for
+/// the mpsc channel or this handler, it would have not dealt with dropped futures.
+/// Weather the added complexity of this design is justified by being able to handle dropped
+/// futures is not clear to me (RaduW) at this moment.
+impl StreamHandler<TrackedFutureFinished, ()> for UpstreamRelay {
```

Comment:
This is up for discussion: I think it would be simpler to implement this as a plain message with `context.notify()` for the following reasons:
1. you don't need to handle and worry about stream errors
2. you don't need to keep the stream around in the upstream actor struct
3. it aligns very well with other handlers in the upstream actor

Reply:
done